api/src/main/java/io/grpc/NameResolver.java

```diff
@@ -568,6 +597,16 @@ public Builder setServiceConfigParser(ServiceConfigParser parser) {
         return this;
       }
 
+      /**
+       * See {@link Args#getChannelLogger}. This is an optional field.
```

Comment:
Why does this need to be optional?

Reply:
I feel like the requirement is too strong for user that must provide a `ChannelLogger`. So if it is not set, it will just default to a no-op channel logger. Do you think it's better to make this required?

Reply:
gRPC is the provider of the Args in most, if not all of the cases. Users of this API are on the consumer side, and they will appreciate a guaranteed non-null ChannelLogger, rather than having to check for null every time they use it.

Reply:
I don't think we need to call out it is an optional field. We will always set it and if their test doesn't they will just get a NPE. I don't think we should have validation in the builder to guarantee it is there, but that doesn't mean consumers of it need to check for nullness.

Reply:
You two have different opinions... So should we make this required in API?

I would prefer this being required (given that we do not default to no-op channel logger if not set) and provides a non-null guarantee to users. In general, I don't like `@Nullable` APIs and would like to avoid as much as possible.

Reply:
As discussed offline, we are trying to make a "soft requirement" for `ChannelLogger` in `NameResolver.Args`. That is, it is only required for `NameResolver.Args` users that `ChannelLogger` matters to them.